matrices.go

```diff
@@ -135,132 +133,142 @@ func (rg reqGraph) CodeOrderInfo() (info OrderInfo) {
 	for _, r := range reqsLow {
 		count := 0
 		for _, parentReq := range r.Parents {
 			if parentReq.Level == to {
-				row := [2]*MatrixItem{NewReqMatrixItem(r), NewReqMatrixItem(parentReq)}
+				row := TableRow{newReqTableCell(r), newReqTableCell(parentReq)}
 				items = append(items, row)
 				count++
 			}
 		}
 		if count == 0 {
-			row := [2]*MatrixItem{NewReqMatrixItem(r), nil}
+			row := TableRow{newReqTableCell(r), nil}
 			items = append(items, row)
 		}
 	}
 	return items
 }
 
-// createSWLCodeMatrix creates a downstream matrix mapping
-// low level requirements to code procedures.
-func (rg *reqGraph) createSWLCodeMatrix() [][2]*MatrixItem {
-	reqs := rg.reqsOfLevel(config.LOW)
-	items := make([][2]*MatrixItem, 0, len(reqs))
-	for _, r := range reqs {
-		count := 0
-		for _, codeTag := range r.Tags {
-			row := [2]*MatrixItem{NewReqMatrixItem(r), NewCodeMatrixItem(codeTag)}
-			items = append(items, row)
-			count++
-		}
-		if count == 0 {
-			row := [2]*MatrixItem{NewReqMatrixItem(r), nil}
-			items = append(items, row)
+// reqsOfLevel returns the non-deleted requirements of the specified level, mapped by ID.
+// @llr REQ-TRAQ-SWL-14
+func (rg ReqGraph) reqsOfLevel(level config.RequirementLevel) map[string]*Req {
+	reqs := make(map[string]*Req, 0)
+	for _, r := range rg.Reqs {
+		if r.Level == level && !r.IsDeleted() {
+			reqs[r.ID] = r
 		}
 	}
-	return items
+	return reqs
 }
 
-// createCodeSWLMatrix creates an upstream matrix mapping
-// code procedures to low level requirements.
-func (rg *reqGraph) createCodeSWLMatrix() [][2]*MatrixItem {
-
-	items := make([][2]*MatrixItem, 0)
-	for _, tags := range rg.CodeTags {
-		for _, codeTag := range tags {
-			count := 0
-			for _, parentReq := range codeTag.Parents {
-				row := [2]*MatrixItem{NewCodeMatrixItem(codeTag), NewReqMatrixItem(parentReq)}
-				items = append(items, row)
-				count++
-			}
-			if count == 0 {
-				row := [2]*MatrixItem{NewCodeMatrixItem(codeTag), nil}
-				items = append(items, row)
+// sortMatrices prepares the sort info and sorts the specified matrices.
+// @llr REQ-TRAQ-SWL-42, REQ-TRAQ-SWL-43, REQ-TRAQ-SWL-44
+func (rg ReqGraph) sortMatrices(matrices ...[]TableRow) {
+	codeOrderInfo := rg.codeOrderInfo()
+	for _, matrix := range matrices {
+		// Update each item's OrderNumber.
+		for _, row := range matrix {
+			for _, item := range row {
+				if item != nil {
+					// Updated order number
+					if item.req != nil {
+						// requirements sorted by ID number
+						item.OrderNumber = item.req.IDNumber
+					} else if item.code != nil {
+						// code sorted by filename and then line number of function
+						item.OrderNumber = codeOrderInfo.filesIndex[item.code.Path]*codeOrderInfo.fileIndexFactor + item.code.Line
+					} else {
+						item.OrderNumber = 0
+					}
+				}
 			}
 		}
+		// Sorts the rows based on the OrderNumber of the items.
+		sort.Slice(matrix, func(i, j int) bool {
+			a0, b0 := matrix[i][0], matrix[j][0]
```

Comment:
I just realized that a0, b0, a1 or b1 can be a nullptr, so we need to deal with that situation here too.

Reply:
hmm, good point. but on closer inspection I don't think it can happen. nullptr can only be in second cell of row (so only a1 & b1), and the code to check the second cell only triggers if the first cell is the same. so, we would need two table rows with the same first cell and nil in the second, which can't happen.

Reply:
You're right this should not happen, but I did see it happening so I looked into it further now. The problem is that some cannot code lines cannot be ordered because 2 tags are in the same line and file:
```
```

Reply:
Probably we should just filter anonymous functions tbh...

Reply:
Let's handle this in a separate PR